### PR TITLE
txn manager status

### PR DESCRIPTION
**Pull request by nziebart**
**Goals (and why)**:
Expose a status on the transaction manager, since it may not be initialized immediately.

Fixes https://github.com/palantir/atlasdb/issues/2440

**Implementation Description (bullets)**:
- Add an `isInitialized()` method and document it
- When asked for the current status, attempt to initialize if not

**Concerns (what feedback would you like?)**:
@gmaretic @fsamuel-bs I'm not sure if this is the best approach or if you guys had already thought of a better way of doing this. Would it be better to extend AsyncInitializer?

TODO: test
**Where should we start reviewing?**:
one file

**Priority (whenever / two weeks / yesterday)**:

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2441)

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java [hunk 1]
```diff
@@ -69,6 +66,31 @@ public LockService getLockService() {
...
+        @Override
+        public boolean isInitialized() {
+            if (!isInitialized) {
+                tryInitialize();
+            }
+
+            return isInitialized;
+        }
+
+        private synchronized void tryInitialize() {
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java]: Rename to `checkInitialized`? We are not actually doing the initialization here, just checking if it was done in the background.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java [hunk 2]
```diff
@@ -69,6 +66,31 @@ public LockService getLockService() {
...
+                return;
+            }
+
+            try {
+                // SerializableTransactionManager should throw until the underlying KVS is initialized.
+                // TODO(ssouza): replace with KVS healthcheck status when that gets implemented.
+                manager.getKeyValueService().getClusterAvailabilityStatus();
+                isInitialized = true;
+            } catch (NotInitializedException e) {
+                log.info("The KeyValueService is not initialized yet!");
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java]: Do we even want to log here when a healthcheck fails? I'd only log in the delegate when `isInitialized()` returns false.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java [hunk 3]
```diff
@@ -39,24 +40,20 @@
     public static class InitializeCheckingWrapper extends AutoDelegate_SerializableTransactionManager {
         private SerializableTransactionManager manager;
         private volatile boolean isInitialized = false;
+        private volatile NotInitializedException initializationException;
+
+        private final RateLimiter tryInitializeLimiter = RateLimiter.create(0.2); // limit 1 per 5 sec
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java]: This is sweet, also deals nicely with excessive spamming of `delegate()` 👍

**Comment by gmaretic** [comment]
I think this is cleaner than extending `AsyncInitializer`

**Comment by fsamuel-bs** [comment]
@nziebart let's chat about this today, since I've had some ideas on how to do this better, since I think the current healthcheck is not sufficient to achieve what you want in this PR.

**Comment by nziebart** [comment]
Talked to @fsamuel-bs offline. This solution is not robust, as the current check for the kvs being ready is not sufficient to execute transactions; we also need other things such as the `TransactionManagersInitializer` (and eventually timelock to have been contacted) to have initialized

Going to close this in favor of creating a queue/list of all dependencies that need to be ready and checking all of them

**Comment by gsheasby** [comment]
Replaced by #2476, which has merged and was released in 0.60.0

**Pull request closed without merging**